pkg/guestagent/docker/events.go

```diff
@@ -0,0 +1,228 @@
+package docker
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/events"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/lima-vm/lima/pkg/guestagent/api"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	startEvent = "start"
+	stopEvent  = "stop"
+	// die event is a confirmation of kill event.
+	dieEvent = "die"
+)
+
+type EventMonitor struct {
+	dockerClient *client.Client
+	// We maintain a record of all active containers because neither the stop nor
+	// die events provide the port mapping. As API consumers, it's our responsibility
+	// to track this information. The map uses the container ID as the key and
+	// stores all published ports associated with that container as the value.
+	runningContainers    map[string][]*api.IPPort
+	runningContainersMux sync.Mutex
+}
+
+func NewEventMonitor() *EventMonitor {
+	return &EventMonitor{
+		runningContainers: make(map[string][]*api.IPPort),
+	}
+}
+
+func (e *EventMonitor) MonitorPorts(ctx context.Context, ch chan *api.Event) {
+	e.tryGetClient(ctx)
```

Comment:
Needs error checking in case the context gets cancelled:

```suggestion
	if err := e.tryGetClient(ctx); err != nil {
		logrus.Errorf("...: %v", err)
		return
	}
```